File: test/ui/charts/bar_label_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:openvitals/ui/charts/bar_chart.dart';
import 'package:openvitals/ui/theme/app_theme.dart';

/// The value label on a bar.
///
/// It used to be measured against the BAR and dropped whole when it did not
/// fit. A bar is its slot minus the gap either side of it, so the label had
/// about 35 logical pixels to live in — and "21,104" needs about 40 where
/// "9,785" needs 34. One character was the difference between a number and
/// nothing at all, which is why every day over 10,000 steps came out blank, and
/// why it was always the tallest bar on the chart that said nothing.
///
/// NOTE ON THE NUMBERS BELOW. Widget tests use a placeholder font whose every
/// glyph is a square of the font size, so text here is far wider than the same
/// text on a device (six digits measure 69px at 11pt in a test, ~40px on a
/// phone). The `maxWidth` values are therefore chosen to exercise the *rule* —
/// fits / shrinks / dropped — and are not device measurements.
void main() {
  late TextStyle style;

  Future<void> withAppTheme(WidgetTester tester) async {
    await tester.pumpWidget(MaterialApp(
      theme: AppTheme.themeFrom(AppTheme.darkScheme),
      home: Builder(builder: (context) {
        style = Theme.of(context).textTheme.labelSmall!;
        return const SizedBox();
      }),
    ));
  }

  BarLabelLayout? layout(String text, double maxWidth) => layoutBarLabel(
        text: text,
        style: style,
        maxWidth: maxWidth,
        textDirection: TextDirection.ltr,
      );

  testWidgets('a label that fits is drawn at full size', (tester) async {
    await withAppTheme(tester);

    final result = layout('21,104 steps', 80)!;

    expect(result.lines.first.text!.style!.fontSize, style.fontSize);
  });

  testWidgets('a label that does not fit SHRINKS — it does not vanish',
      (tester) async {
    await withAppTheme(tester);

    // Too narrow for the number at 11pt, wide enough at a smaller size. This is
    // the case that used to return null and leave the bar blank.
    final result = layout('21,104 steps', 55);

    expect(result, isNotNull);
    expect(result!.width, lessThanOrEqualTo(55));
    expect(
      result.lines.first.text!.style!.fontSize,
      lessThan(style.fontSize!),
      reason: 'it should have stepped down to fit, not given up',
    );
  });

  testWidgets('the longer the number, the smaller it is drawn — but it is drawn',
      (tester) async {
    await withAppTheme(tester);

    final shorter = layout('9,785 steps', 60)!;
    final longer = layout('218,104 steps', 60)!;

    expect(longer.width, lessThanOrEqualTo(60));
    expect(
      longer.lines.first.text!.style!.fontSize,
      lessThan(shorter.lines.first.text!.style!.fontSize!),
    );
  });

  testWidgets('the unit goes on its own line, so the number keeps the room',
      (tester) async {
    await withAppTheme(tester);

    expect(splitBarLabel('21,104 steps'), ['21,104', 'steps']);
    expect(layout('21,104 steps', 80)!.lines, hasLength(2));
    // Nothing to split: one line.
    expect(splitBarLabel('21,104'), isNull);
  });

  testWidgets('a label nobody could read is still dropped', (tester) async {
    await withAppTheme(tester);

    // A month chart gives 31 bars a few pixels each. Below the floor a number
    // is not worth drawing, and drawing it anyway would read worse than the gap.
    expect(layout('21,104 steps', 10), isNull);
    expect(layout('    ', 100), isNull);
    expect(layout('21,104 steps', 0), isNull);
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/ui/charts/BarLabelTest.kt
package tech.mmarca.openvitals.ui.components

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The pure half of the bar value label. Kotlin measures at one size and drops on overflow,
 * so the Dart shrink-to-fit cases are not portable. The fake measurer is six px per character, ten px a line.
 */
class BarLabelTest {

    /** Measures with the fake measurer: six px per character, ten px a line. */
    private fun measure(text: String, maxWidth: Float): BarLabelMeasurement<String>? =
        barLabelLines(text)?.let { lines ->
            measureBarLabelLines(
                lines = lines,
                maxWidth = maxWidth,
                lineGap = 1,
                measure = { it },
                lineWidth = { it.length * 6 },
                lineHeight = { 10 },
            )
        }

    @Test
    fun `a label that fits is drawn at full size`() {
        val result = measure("21,104 steps", maxWidth = 80f)

        // Kotlin measures at one size, so this is the fits-therefore-drawn half of the rule.
        assertNotNull(result)
        assertTrue(result!!.width <= 80)
        assertEquals(2, result.lines.size)
    }

    @Test
    fun `the unit goes on its own line, so the number keeps the room`() {
        // Kotlin strips the grouping separators, a deliberate divergence from Dart.
        assertEquals(listOf("21104", "steps"), splitBarValueLabel("21,104 steps"))
        assertEquals(2, measure("21,104 steps", maxWidth = 80f)!!.lines.size)
        // Nothing to split: one line.
        assertNull(splitBarValueLabel("21,104"))
        assertEquals(1, measure("21,104", maxWidth = 80f)!!.lines.size)
    }

    @Test
    fun `a label nobody could read is still dropped`() {
        // A month chart gives 31 bars a few pixels each. A label that overruns its slot is dropped.
        assertNull(measure("21,104 steps", maxWidth = 10f))
        assertNull(measure("    ", maxWidth = 100f))
        assertNull(measure("21,104 steps", maxWidth = 0f))
    }

    @Test
    fun `the measured label reports the widest line and the stacked height`() {
        val result = measure("21,104 steps", maxWidth = 80f)!!

        // Both lines are five characters: 30 px wide, two 10 px lines with a 1 px gap.
        assertEquals(30, result.width)
        assertEquals(21, result.height)
    }
}
